### DISCUSSION TITLE
Single instance as argument to multiple constructors

### DISCUSSION THREAD

**Discussion by jessegeens** [LinkedSoftwareDependencies/Components.js#75] [Q&A] [answered]
Hey all,
Sorry if this is a somewhat dumb question. I'm trying to extend the Solid Community Server, which uses components.js as its DI framework. 

I have a component (`AnonymizingHttpHandler`) that takes multiple instances of different classes as its arguments. One of these is a component called the `ConfigurationManager`, which stores all the user configuration details. Some of its other arguments are instances of different classes, but these instances also take as an argument in their constructor an instance of this `ConfigurationManager`. 

I'm trying to make sure that only one instance of the `ConfigurationManager` is created (since it reads a number of config files). I'm not sure if this happens the way I have my configuration specified right now; and if it doesn't I don't really have an idea on how to configure this. I would appreciate it if anyone could help me out!

My config file is:
```json
"args_operationHandler": {
          "@type": "AnonymizingHttpHandler",
          "args_credentialsExtractor": { "@id": "urn:solid-server:default:CredentialsExtractor" },
          "args_modesExtractor": { "@id": "urn:solid-server:default:ModesExtractor" },
          "args_permissionReader": { "@id": "urn:solid-server:default:PermissionReader" },
          "args_authorizer": { "@id": "urn:solid-server:default:Authorizer" },
          "args_operationHandler": { "@id": "urn:solid-server:default:OperationHandler" },
          "args_dataTreatmentHandler": {
            "@id": "urn:pepsa-component:default:DataTreatmentHandler",
            "@type": "DataTreatmentHandler",
            "configMgr": { "@id": "urn:pepsa-component:default:ConfigurationManager" },
            "ruleEncapsulator": {
              "@id": "urn:pepsa-component:default:RuleEncapsulator",
              "@type": "RuleEncapsulator",
              "configMgr": { "@id": "urn:pepsa-component:default:ConfigurationManager" }
            },
            "parserSelector": {
              "@id": "urn:pepsa-component:default:ParserSelector",
              "@type": "ParserSelector"
            }
          },
          "args_configurationManager": {
            "@id": "urn:pepsa-component:default:ConfigurationManager",
            "@type": "ConfigurationManager"
          }
}
``` 

Thanks a lot!

**Comment by joachimvh** [comment, marked as answer]
This will indeed only create 1 instance of your `ConfigurationManager`. You could put a `console.log` call in the constructor to see how many times it is called. Identical `@id`s always corresponds to the same instance of a class. Components.js reads out all the JSON-LD files and interprets them as RDF, so from its point of view there is just a single `urn:pepsa-component:default:ConfigurationManager` subject that has the `ConfigurationManager` type, and some other triples referencing that same URI.

The wrong way to do it would have been to replace those `{ "@id": "urn:pepsa-component:default:ConfigurationManager" }` blocks with `{ "@type": "ConfigurationManager" }`. This would have created several blank nodes that each had the `ConfigurationManager` type.

**Reply by jessegeens** [reply]
Perfect, thanks for the very clear answer!!
